### PR TITLE
[V3 Audio] Add Spotify support

### PR DESCRIPTION
**Pull request by aikaterna**
### Type

- [ ] Bugfix
- [ ] Enhancement
- [x] New feature

### Description of the changes
This adds the ability to use open.spotify.com links or spotify:track:code type of Spotify tracks with the play command. Tracks, albums, playlists, and user playlists can be added using the url or code, but all within the play command. This requires that the bot owner sign up for a Spotify API key along with a YouTube Data API v3 key, and input the Spotify ID, Spotify client secret, and the YouTube API key in [p]audioset commands before those URLs/codes can be used. Thanks to TrustyJAID for his help in moving the YouTube API access to something quicker and more efficient than what I originally had in mind.

### CODE REVIEW

**Comment by palmtree5** [comment]
it might be worth reworking some things to deduplicate code as Streams also uses the Youtube Data API V3 for requests related to youtube livestreams and has its own command to set the api key

**Comment by mikeshardmind** [comment]
Needs some small conflict handling. Changes provided are nice. I think there needs to be a few extra things included to handle limiting exceptionally large playlists later on, but that can easily be a further improvement, rather than something which needs to be included with this specifically.

I agree with @palmtree5 about a more central handling of secrets. I think we should look to store these in the bot's central config instead of per cog (with these and any other API keys used by core red) but this can also be done seperately to this.

**Comment by Tobotimus** [comment]
This'd be great for 3.1. I'm going to set #2389 as a blocker for this one due to the discussion about central storage of API keys.

**Review by TrustyJAID — DISMISSED** [review]
This has been working great in my testing. I would like to see the bot start playing some tracks immediately instead of waiting until all the playlist songs are found before starting but that can be figured out in a later PR.

File: redbot/cogs/audio/audio.py [hunk 1]
```diff
@@ -438,6 +441,33 @@ def _restart_connect(self):
...
+            f"1. Go to Spotify developers and log in with your Spotify account\n"
+            "(https://developer.spotify.com/dashboard/applications)\n"
+            '2. Click "Create An App"\n'
+            "3. Fill out the form provided with your app name, etc.\n"
+            '4. When asked if you\'re developing commercial integration select "No"\n'
+            "5. Accept the terms and conditions.\n"
+            "6. Copy your client ID and your client secret into\n"
+            "`{prefix}set api spotify client_id,your_client_id "
+            "client_secret,your_client_secret`".format(prefix=ctx.prefix)
+        )
```
Comment (Tobotimus) [inline redbot/cogs/audio/audio.py]: `.format` here needs to be outside the `_()`

File: redbot/cogs/audio/audio.py [hunk 2]
```diff
@@ -470,21 +500,32 @@ def _restart_connect(self):
...
+            "2. You should be prompted to create a new project (name does not matter).\n"
+            "3. Click on Enable APIs and Services at the top.\n"
+            "4. In the list of APIs choose or search for YouTube Data API v3 and click on it. Choose Enable.\n"
+            "5. Click on Credentials on the left navigation bar.\n"
+            "6. Click on Create Credential at the top.\n"
+            '7. At the top click the link for "API key".\n'
+            "8. No application restrictions are needed. Click Create at the bottom.\n"
+            "9. You now have a key to add to `{prefix}set api youtube api_key,your_api_key`".format(
+                prefix=ctx.prefix
+            )
```
Comment (Tobotimus) [inline redbot/cogs/audio/audio.py]: `.format` here needs to be outside the `_()`

File: redbot/cogs/audio/audio.py [hunk 3]
```diff
@@ -1082,6 +1245,131 @@ def _restart_connect(self):
...
+
+        if len(track_list) == 0:
+            embed3 = discord.Embed(
+                colour=await ctx.embed_colour(),
+                title=_(
+                    "Nothing found.\nThe YouTube API key may be invalid "
+                    "or you may be rate limited on YouTube's search service.\n"
+                    "Check the YouTube API key again and follow the instructions "
+                    "at `{prefix}audioset youtubeapi`.".format(prefix=ctx.prefix)
+                ),
```
Comment (Tobotimus) [inline redbot/cogs/audio/audio.py]: `.format` here needs to be outside the `_()`

**Review by Tobotimus — DISMISSED** [review]
Just need some easy fixes around the i18n strings. I think there is a 4th one earlier on in audio.py which is not a result of these changes, but I'm happy for you to fix it up here anyway.

**Review by tekulvw — APPROVED** [review]
Changes requested by Toby have been made

**Pull request merged**